src/chapter5/message-passing.md

```diff
@@ -1 +1,240 @@
+
+### Point-to-Point communication
+
+These are blocking functions - they wait until the message is sent or received. Note that the CPU is actively polling the network interface when waiting for a message. This is opposite in behaviour to other C functions, i.e. c= getChar() (which causes a context switch and then a sleep in the OS). This is done for speed reasons.
+
+```C
+int MPI_Send(void * buf, int count, MPI_Datatype type, int dest, int tag, MPI_Comm comm);
+```
+
+Sends a message from the calling process to another process
+
+INPUT PARAMETERS
+
+- buf
+  - Initial address of send buffer (choice).
+- count
+  - Number of elements sent (non negative integer).
+- type
+  - DataType of each send buffer element (handle).
+- dest
+  - Rank of destination (integer).
+- tag
+  - Message tag (integer).
+- comm
+  - Communicator (handle).
+
+OUTPUT PARAMETER
+
+- IERROR
+  - Fortran only: Error status (integer).
+
+```c
+int MPI_Recv(void * buf, int count, MPI_Datatype type, int source, int tag, MPI_Comm comm, MPI_Status * status);
+```
+
+Receives a message from another process
+
+INPUT PARAMETERS
+
+- count
+  - Maximum number of elements to receive (integer).
+- type
+  - DataType of each receive buffer entry (handle).
+- source
+  - Rank of source (integer).
+- tag
+  - Message tag (integer).
+- comm
+  - Communicator (handle).
+
+OUTPUT PARAMETERS
+
+- buf
+  - Initial address of receive buffer (choice).
+- status
+  - Status object (status).
+- IERROR
+  - Fortran only: Error status (integer).
+
+### Primary MPI Routines closing
+
+In a header file you will find
+
+``` C
+int MPI_Finalize(void);
+```
+
+To call in your C/C++ program
+
+``` C
+#include <mpi.h>
+MPI_Finalize();
+```
+
+## General overview MPI program
+
+``` C
+...
+int MPI_Init(int  argc, char ** argv);
+--------------------------Parallel algorithm starts----------------------
+int MPI_Comm_size(MPI_Comm comm, int * size);
+int MPI_Comm_rank(MPI_Comm comm, int * rank);
+...
+int MPI_Send(void * buf, int count, MPI_Datatype type, int dest, int tag, MPI_Comm comm);
+int MPI_Recv(void * buf, int count, MPI_Datatype type, int source, int tag, MPI_Comm comm, MPI_Status * status);
+…
+--------------------------Parallel algorithm ends-----------------------
+int MPI_Finalize(void);
+...
+
+```
+
+Use man pages to find out more about each routine
+
+When sending a Process it packs up all of its necessary data into a buffer for the receiving process. These buffers are often referred to as envelopes since the data is being packed into a single message before transmission (similar to how letters are packed into envelopes before transmission to the post office)
+
+## Elementary MPI Data types
+
+MPI_Send and MPI_Recv utilize MPI Datatypes as a means to specify the structure of a message at a higher level. The data types defined in the table below are simple in nature and for custom data structures you will have to define the structure.
+
+![MPI datatypes](../imgs/mpi_datatypes.png)
```

Comment:
Maybe create your own table in Markdown. [Markdown table generator](https://www.tablesgenerator.com/markdown_tables)